### ISSUE TITLE
TextField on desktop with lots of text can hang app

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#90063]
For example, this app:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

void main() {
  final TextEditingController controller = TextEditingController();
  runApp(
    MaterialApp(
      home: Material(
        child: Center(
          child: Column(
            children: [
              TextField(controller: controller),
              TextButton(
                onPressed: () {
                  controller.text = 'abcdef' * 100000;
                },
                child: Text('Fill'),
              ),
            ],
          ),
        ),
      ),
    ),
  );
}
```

on my MacBook pro causes the app to freeze after pressing fill. I discovered this when accidentally pasting a very large amount of text into a text field while testing something else. In that case, the text field probably also had some extra styling and input formatters, and the app completely hung (and hot restarting caused it to crash). This sample app does not crash on hot restart, but I suspect adding formatters or making the text even larger might cause it to.

**Comment by maheshmnj** [comment]
I am able to reproduce the issue on m1 mac the app is almost frozen, verified on `stable 2.5.1` and `master 2.6.0-6.0.pre.171`. I could not reproduce the crash (on hot restart) after adding `style`. 

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  final TextEditingController controller = TextEditingController();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
            TextField(
              controller: controller,
              style: TextStyle(
                  fontSize: 20,
                  color: Colors.red,
                  letterSpacing: 2,
                  fontStyle: FontStyle.italic),
              maxLines: 5,
            ),
            TextButton(
              onPressed: () {
                controller.text = 'abcdef' * 100000;
              },
              child: Text('Fill'),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

```
</details>




<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.5.1, on macOS 11.6 20G165 darwin-arm, locale en-GB)
    • Flutter version 2.5.1 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (5 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.59.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.26.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.6 20G165 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 94.0.4606.54

• No issues found!
```

```
[✓] Flutter (Channel master, 2.6.0-6.0.pre.171, on macOS 11.6 20G165 darwin-arm, locale en-GB)
    • Flutter version 2.6.0-6.0.pre.171 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 321c872ca7 (6 hours ago), 2021-09-22 02:38:06 -0400
    • Engine revision b0f3c0f7e4
    • Dart version 2.15.0 (build 2.15.0-127.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.5.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.59.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.26.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.6 20G165 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 94.0.4606.54

• No issues found!
```
</details>


on a side note: adding `maxlines` to textfield it takes much longer (around 10s) to fill text into textfied.

**Comment by justinmc** [comment]
Thanks for the reproducible issue, this has been in the back of my mind for awhile.  I want to figure out exactly what the bottleneck is here.  Just some quick thoughts for now:

  * This isn't really an insurmountable amount of text (~800kb back of the napkin?).
  * Currently, the entire contents of the field is synced between the framework and engine on every change, which is going to be a performance concern even if it's not exactly the cause of this slowdown here.
  * @Renzo-Olivares introduced the TextEditingDeltas API in https://github.com/flutter/flutter/pull/88477, which may give us a way to work with deltas instead of the entire string.

**Comment by cbracken** [comment]
flutter/engine#29036 may help for macOS.

**Comment by letungcntt** [comment]
@cbracken Has this problem been solved ?

**Comment by cbracken** [comment]
@letungcntt -- @justinmc will be working on migrating Windows to a delta-based approach but this has not yet landed.

**Comment by letungcntt** [comment]
> [flutter/engine#29036](https://github.com/flutter/engine/pull/29036) may help for macOS.

i has checkout to beta 2.9.0 and change enableDeltaModel = TRUE but it isn't working. How to use TextEditingDelta ?
<platform: MacOS>

**Comment by justinmc** [comment]
@letungcntt This is the PR you'll need: https://github.com/flutter/engine/pull/29036.  It was rolled into the framework in https://github.com/flutter/flutter/pull/93941 in 2.9.0-0.1.pre, but it might have been reverted and rolled in again later...  Anyway it seems like your version 2.9.0 is slightly too old.  Master definitely has it.

Also, I have to point out that this is all still early and experimental and not currently a viable fix for this issue.  Using `enableDeltaModel = true` will completely break EditableText and TextField, etc.  Currently this feature is really only for those who are implementing their own low-level text editors.  Lastly, we currently send the old state of the field along with the delta on each change ([oldText docs](https://master-api.flutter.dev/flutter/services/TextEditingDelta/oldText.html)), so there is no memory improvement unless we remove this.

Tldr; It's very early for this feature.  You won't be able to improve performance for large amounts of text unless you are ready to really hack text editing on your own, and even then nothing is guaranteed to improve.

**Comment by PaulEibensteiner** [comment]
Hi Everybody, as I understand the PR has since been merged. Still I can not get it to work on windows, even when setting the option `enableDeltaModel: true` I still only receive calls to `updateEditingValue`. Should this be working? I can post an mwp here if you want.

**Comment by Renzo-Olivares** [comment]
@PaulEibensteiner Are you doing this while using your own implementation of `DeltaTextInputClient` or are you using Flutter's included `EditableText`/`TextField`? If it's the latter you would have to change `implements TextInputClient` to `implements DeltaTextInputClient` and define `updateEditingValueWithDeltas` accordingly. Currently Flutter does not provide a `DeltaTextInputClient` implementation. However this will not fix the performance issue because deltas are only sent one way from the `engine` to the `framework` and not vice versa. There is still more work to be done around this issue.

**Comment by PaulEibensteiner** [comment]
I am using my own implementation, something along the lines of [this](https://gist.github.com/PaulEibensteiner/d934f1751ab0c4ba2ee0a31b1c9c69e5), which does not work for me as explained above.

**Comment by PaulEibensteiner** [comment]
Btw, for a custom more performant text field, do you think it would be possible to simply set the current text editing value only to the paragraph(s) currently being edited via `TextInputConnection.setEditingState`? It seems to me that this should prevent the TextInputConnection from sending the whole document back and forth.

**Comment by Renzo-Olivares** [comment]
What platform/flutter version are you running on? That sounds like it could be part of a solution, though this must also be handled on the `TextInputPlugin` for the platform on the engine side. Also there is still work to be done on the performance of laying out large amounts of text. See: #92173

**Comment by PaulEibensteiner** [comment]
Flutter 2.10.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision c860cba910 (3 weeks ago) • 2022-03-25 00:23:12 -0500
Engine • revision 57d3bac3dd
Tools • Dart 2.16.2 • DevTools 2.9.2

sorry forgot about that

**Comment by Renzo-Olivares** [comment]
I think the engine that comes with Flutter 2.10.4 stable doesn't yet support deltas on all platforms. The master branch currently supports all platforms.

**Comment by MegatronKing** [comment]
TextInput.setEditingState for a large text is still expensive, will have a delta api in the future?

**Comment by Renzo-Olivares** [comment]
@cbracken For all desktop platforms deltas are currently supported but the `TextField` has not been migrated over to use it, you would have to create your own `DeltaTextInputClient` to use it. We currently do not support sending deltas to the platforms text input plugin through `TextInput.setEditingState`.

**Comment by dnfield** [comment]
I don't think we can safely drop the JSON codec without some major ecosystem migration. A specific one for text input would be good if we don't have it.

It might also be good to have an umbrella issue to drop the JSON codec usage within the framework, although I think some of the usages that are left outside of text are channels that tend to have small and rarely used messages.

**Comment by MegatronKing** [comment]
@Renzo-Olivares Thanks for `TextInputClient/DeltaTextInputClient`, it solved my problem. I believe`TextInput.setEditingState` should have a delta api, which can significantly improve performance on a large text. A basic design, selection and text should be separated when sending to platforms. The current situation is that any change will send both of them to platforms, even if I just changed the selection offset.

https://user-images.githubusercontent.com/15059084/175192973-a1458bf1-0f7e-456a-ba3e-1deaab733521.mov

**Comment by huycozy** [comment]
Hi guys, I also can reproduce this issue on mobile and web platforms (on the latest `stable` and `master` channels) with sample code at https://github.com/flutter/flutter/issues/90063#issuecomment-924881999 and https://github.com/flutter/flutter/issues/114158#issue-1425765173. 

This issue is being discussed but I see it being added to the Desktop project. I think we should consider updating appropriate labels and project if any.

Related issues/discussion: 

- https://github.com/flutter/flutter/issues/107470#issuecomment-1188647524
- https://github.com/flutter/flutter/issues/114158

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.3.6, on macOS 12.6 21G115 darwin-x64, locale en-VN)
    • Flutter version 3.3.6 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6928314d50 (30 hours ago), 2022-10-25 16:34:41 -0400
    • Engine revision 3ad69d7be3
    • Dart version 2.18.2
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode14.0.1.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.72.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • SM T225 (mobile) • R9JT3004VRJ • android-arm64  • Android 12 (API 31)
    • macOS (desktop)  • macos       • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)     • chrome      • web-javascript • Google Chrome 107.0.5304.62

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

```bash
[!] Flutter (Channel master, 3.5.0-10.0.pre.166, on macOS 12.6 21G115 darwin-x64, locale en-VN)
    • Flutter version 3.5.0-10.0.pre.166 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4f3f932a68 (5 hours ago), 2022-10-30 17:44:24 -0400
    • Engine revision f91119f84f
    • Dart version 2.19.0 (build 2.19.0-353.0.dev)
    • DevTools version 2.18.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode14.0.1.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Community Edition (version 2022.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 70.0.5
    • Dart plugin version 222.4167.21

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.72.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 107.0.5304.87

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```
</details>

**Comment by kaboc** [comment]
Another issue #114158 I opened was closed as a duplicate, but one of the two examples there that uses TextSpan may be helpful. I'm not sure if it has the same cause, so please refer to the page to make sure the issue demonstrated with the example is also fixed when the issue in this page is fixed.

**Comment by farshed** [comment]
In case this helps someone pinpoint the exact issue.

For my text editing app, I display a `ListView.builder` of Text lines with custom selection highlights painted on top and it has turned out very smooth. TextField on the other hand, had a noticeable lag while scrolling. Not sure if TextField's content is virtualized?

The other lag happens while editing the text which could be due to the underlying text buffer (I'm assuming it's just a long string in TextField's case?) Representing the entire document as a single string/list means that the entire sequence would have to be copied or moved in memory on each update which can get very expensive.

**Comment by aleripe** [comment]
> In case this helps someone pinpoint the exact issue.
> 
> For my text editing app, I display a `ListView.builder` of Text lines with custom selection highlights painted on top and it has turned out very smooth. TextField on the other hand, had a noticeable lag while scrolling. Not sure if TextField's content is virtualized?
> 
> The other lag happens while editing the text which could be due to the underlying text buffer (I'm assuming it's just a long string in TextField's case?) Representing the entire document as a single string/list means that the entire sequence would have to be copied or moved in memory on each update which can get very expensive.

Hi, @farshed do you have any sample of your custom selection implementation? Thanks!

**Comment by gnprice** [comment]
Retitled and adjusted labels because in addition to the original report for desktop, several subsequent reports (https://github.com/flutter/flutter/issues/90063#issuecomment-1296495185, #121683) are for web and mobile.

**Comment by Justus-M** [comment]
I have this issue for web. I have users pasting large amounts of text in, and the app hangs for a long time (sometimes just freezes completely)

I've had to use a workaround where I truncate the text and ask to copy it directly from the clipboard, but it's a terrible UX because the browser asks for permission to access their clipboard.

**Comment by Srimathi622** [comment]
> @cbracken For all desktop platforms deltas are currently supported but the `TextField` has not been migrated over to use it, you would have to create your own `DeltaTextInputClient` to use it.
> 
> We have documentation on how to enable it here https://api.flutter.dev/flutter/services/DeltaTextInputClient-class.html . Though we don't have a concrete guide on how to implement your own `TextInputClient/DeltaTextInputClient`.
> 
> We currently do not support sending deltas to the platforms text input plugin through `TextInput.setEditingState`, but this is something we can definitely consider if it will help performance.

is there any document available for creating our own TextField using DeltaTextInputClient? I am facing a performance issue. The speed is reduced when we have a huge record

**Comment by dtruong27** [comment]
We had a workaround but only for the Web platform. We wanted to use a `TextAreaElement` but there is a bug that `onKeyUp` and `onKeyDown` events are not triggered. We then ended up using `IFrameElement` with `<textarea>` in the `srcdoc`. Our workaround looks like this:

```
  html.HtmlElement longtextForm(int id, double height) {
    final iframe = html.IFrameElement()
      ..style.cssText =
          "height:100%;width:100%;scrolling:no;border:1px solid black;background:#FFFFFF;";

    iframe.srcdoc = '''
<script>
  const changeText = function (e) {
    const v = document.getElementById("text-editor").value;
    window.parent.postMessage(v, "*");
  };
</script>
<textarea id="text-editor" style="width:100%;height:100%;min-height:${(height * 0.95).round()}px;
overflow:auto;font-size:11pt;font-weight:300;font-family:Helvetica,sans-serif;resize:none;border:none;outline:none;background:none;margin:0px;box-sizing:border-box;" 
onkeyup=changeText()>
$content
</textarea>
''';

    final div = html.DivElement()
      ..style.width = "100%"
      ..style.height = "100%"
      ..children = [iframe];

    return div;
  }

  void observeEvents() {
    html.window.onMessage.listen((event) {
      final text = event.data;
      if (text != null && text is String) {
        onContentUpdated(text);
      }
    });
  }
```

**Comment by Srimathi622** [comment]
Thanks for your replies. Is there any workaround for the desktop application (Windows and macOS)? I need a code editor to handle huge text.

**Comment by Srimathi622** [comment]
Thanks a lot. It works

**Comment by Markaleth** [comment]
Just to add to the discussion.,  this issue also happens with plain Text or SelectableText widgets.  Or at least, my assumption is that it's the same root cause at play. 

My current use case is trying to render a very large JSON response to the UI in a SelectableText widget with flutter web. The length of the string i'm passing to the SelectableText / Text widget is between 270k and 870k characters. 

This results in the website hanging for anywhere between 250 ms to close to 1000 ms.

**Comment by Maruf-Elsa** [comment]
I had the same issue when validating email. The reason was Flutter was not able to handle complex regex and was hanging the app when the email was more than 35 characters. I hope the Flutter team takes a note of this and fixes it.

Complex regex hanging the app - `r'^[_A-Za-z0-9-+]+(\.[_A-Za-z0-9-]+)*@[A-Za-z0-9-]+(\.[A-Za-z0-9]+)*(\.[A-Za-z]{2,})$'`
Simple working regex for email- `r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'`
